File: desktop_client/engines/activate_pos_engine.py
"""
Python port of office-scripts/ActivatePOSEngine.ts's main(). See
import_engine.py's module docstring for the duplication rationale. See
ActivatePOSEngine.ts's own file header for the full business-rule rationale
(never reassigns a POS's technician, two mutually-exclusive selection
modes, idempotency) - this file intentionally does not re-explain it, only
mirrors the logic.
"""
from __future__ import annotations

import datetime

from .core_logic import category_rule, norm
from .js_compat import at as _at, num as _num
from .mock_workbook import MockWorkbook


def run(workbook: MockWorkbook) -> str:
    def read_table(sheet_name: str) -> list[list]:
        ws = workbook.get_worksheet(sheet_name)
        rng = ws.get_used_range()
        return rng.get_values() if rng else []

    pos_master = read_table("POS_MASTER")
    if len(pos_master) < 2:
        return "Activate POS Engine: POS_MASTER is empty - nothing to do."
    category_rules_raw = read_table("CATEGORY_RULES")
    activate_list_raw = read_table("POS_ACTIVATE_LIST")
    control = read_table("CONTROL")

    def setting(name: str, fallback: float) -> float:
        for i in range(1, len(control)):
            if norm(str(control[i][0])) == norm(name):
                try:
                    return float(control[i][1])
                except (TypeError, ValueError):
                    return fallback
        return fallback

    activate_count_by_ppt = int(setting("ACTIVATE_COUNT_BY_PPT", 0))

    category_rules_table: list[dict] = []
    for i in range(1, len(category_rules_raw)):
        category_rules_table.append({
            "key": norm(str(category_rules_raw[i][0])),
            "value": norm(str(category_rules_raw[i][1])),
        })

    pm_headers = [str(h) for h in pos_master[0]]

    def pm_idx(name: str) -> int:
        return pm_headers.index(name) if name in pm_headers else -1

    pos_id_col = pm_idx("posId")
    status_col = pm_idx("status")
    category_col = pm_idx("category")
    override_type_col = pm_idx("managerOverrideType")
    ppt_col = pm_idx("ppt")
    notes_col = pm_idx("plannerNotes")

    explicit_ids: set[str] = set()
    for i in range(1, len(activate_list_raw)):
        v = str(_at(activate_list_raw[i], 0) or "").strip()
        if v:
            explicit_ids.add(v)

    pos_ws = workbook.get_worksheet("POS_MASTER")
    today = datetime.date.today().isoformat()
    activated: list[str] = []
    already_active = 0
    skipped_force_exclude: list[str] = []

    def activate_row(row_index: int, row: list, via: str) -> None:
        nonlocal already_active
        current_override = norm(str(_at(row, override_type_col) or ""))
        if current_override == "FORCE_INCLUDE":
            already_active += 1
            return
        pos_ws.get_range_by_indexes(row_index, override_type_col, 1, 1).set_value("FORCE_INCLUDE")
        existing_notes = str(_at(row, notes_col) or "")
        pos_ws.get_range_by_indexes(row_index, notes_col, 1, 1).set_value(
            (existing_notes + " | " if existing_notes else "") + "Hromadně aktivováno" + via + " " + today
        )

    mode = ""
    if explicit_ids:
        mode = "seznam (POS_ACTIVATE_LIST)"
        for i in range(1, len(pos_master)):
            row = pos_master[i]
            pos_id = str(_at(row, pos_id_col))
            if pos_id not in explicit_ids:
                continue
            if str(_at(row, status_col)) != "Active":
                continue
            current_override = norm(str(_at(row, override_type_col) or ""))
            if current_override == "FORCE_EXCLUDE":
                skipped_force_exclude.append(pos_id)
                continue
            activate_row(i, row, "")
            activated.append(pos_id)
    elif activate_count_by_ppt > 0:
        mode = f"prvních {activate_count_by_ppt} podle PPT"
        pool: list[dict] = []
        for i in range(1, len(pos_master)):
            row = pos_master[i]
            if str(_at(row, status_col)) != "Active":
                continue
            current_override = norm(str(_at(row, override_type_col) or ""))
            if current_override in ("FORCE_EXCLUDE", "FORCE_INCLUDE"):
                continue
            category = str(_at(row, category_col))
            rule = category_rule(category_rules_table, norm(category))
            if rule != "EXCLUDE":
                continue
            pool.append({
                "index": i, "posId": str(_at(row, pos_id_col)),
                "ppt": _num(_at(row, ppt_col)) or 0, "row": row,
            })
        pool.sort(key=lambda item: item["ppt"], reverse=True)
        for item in pool[:activate_count_by_ppt]:
            activate_row(item["index"], item["row"], " (PPT)")
            activated.append(item["posId"])
    else:
        return "Activate POS Engine: POS_ACTIVATE_LIST is empty and CONTROL.ACTIVATE_COUNT_BY_PPT is 0 - nothing to activate."

    skipped_note = f": {', '.join(skipped_force_exclude)}" if skipped_force_exclude else ""
    return (
        f"Activate POS Engine: mode = {mode}. "
        f"{len(activated)} POS newly/already set to FORCE_INCLUDE ({already_active} already were), "
        f"{len(skipped_force_exclude)} skipped (explicit FORCE_EXCLUDE always wins){skipped_note}. "
        "Run Planning Engine next to see them enter this week's plan."
    )

File: office-scripts/ActivatePOSEngine.ts
// ============================================================================
// FIELD FORCE OPTIMIZER V11 - ACTIVATE POS ENGINE
// ============================================================================
// Deployable Office Script. Run manually, whenever the manager wants to pull
// specific POS back into planning that CATEGORY_RULES currently excludes
// wholesale (e.g. category "1CD"/"1POSTA" - see CATEGORY_RULES).
//
// Product owner (2026-07-11), after being asked whether this should
// reassign POS ownership: "já nikdy do jejich 'přiřazení'... měnit nechci,
// ja chci mít možnost je přidat jako máme třeba teď vyřazené 1CD, ale chci
// mít možnost určit buďto jaké, nebo prvních 500 nehledě kolik techniků to
// bude" - explicitly NOT a reassignment tool. This never touches
// assignedTechnician. It only sets POS_MASTER.managerOverrideType to
// FORCE_INCLUDE on selected POS - the exact mechanism PlanningEngine.ts
// already has for "manually include a POS a category rule would otherwise
// filter out" (see that file's "FORCE_INCLUDE bypasses Filters entirely"
// comment) - so each activated POS still gets planned for its OWN existing
// assignedTechnician (or managerOverrideTechnician if already set), via the
// normal capacity/priority/hold-back rules. However many technicians those
// POS happen to belong to is exactly how many technicians end up touched -
// no manual per-technician split needed, by design.
//
// TWO SELECTION MODES (mutually exclusive; POS_ACTIVATE_LIST wins if it has
// any rows, otherwise CONTROL.ACTIVATE_COUNT_BY_PPT is used):
//   1. Explicit list - POS IDs pasted into POS_ACTIVATE_LIST (same minimal
//      paste-list pattern as BLACKLIST, just the opposite direction).
//      Product owner: "podle mého seznamu" - the manager's own picks.
//   2. Count by PPT - CONTROL.ACTIVATE_COUNT_BY_PPT (default 0 = disabled).
//      Builds the pool of Active POS whose CATEGORY_RULES rule resolves to
//      EXCLUDE and that aren't already FORCE_INCLUDE/FORCE_EXCLUDE, sorts by
//      ppt descending, activates the top N. Product owner: "nebo ppt".
//
// Idempotent: re-running with the same inputs is safe - a POS already
// FORCE_INCLUDE is left alone (no duplicate note appended), and the count
// mode's pool always excludes anything already activated, so a second run
// with the same N naturally picks up the NEXT-highest-PPT still-excluded
// POS rather than re-doing the same ones.
//
// Does not touch managerOverridePriority or managerOverrideTechnician -
// only managerOverrideType, plus a short audit trail appended to
// plannerNotes (never overwritten, only appended to).
// ============================================================================

function main(workbook: ExcelScript.Workbook) {
  function readTable(sheetName: string): (string | number | boolean)[][] {
    const ws = workbook.getWorksheet(sheetName);
    const range = ws.getUsedRange();
    return range ? range.getValues() : [];
  }

  // SYNC-BLOCK-START: text.ts
  function norm(v: string): string {
    return v
      .toUpperCase()
      .normalize("NFD")
      .replace(/[\u0300-\u036f]/g, "")
      .trim();
  }
  // SYNC-BLOCK-END: text.ts

  // SYNC-BLOCK-START: core.ts (activate)
  function categoryRule(
    categoryRulesTable: { key: string; value: string }[], // key/value already normalized (upper, no diacritics)
    categoryNormalized: string
  ): string {
    let starPrefixRule: string | null = null;
    for (const row of categoryRulesTable) {
      if (row.key == categoryNormalized) {
        return row.value; // exact match always wins immediately
      }
      if (row.key == "STARTS_1" && categoryNormalized.startsWith("1")) {
        starPrefixRule = row.value;
      }
      if (row.key == "*") {
        starPrefixRule = starPrefixRule ?? row.value;
      }
    }
    return starPrefixRule ?? "NORMAL";
  }
  // SYNC-BLOCK-END: core.ts (activate)

  const posMaster = readTable("POS_MASTER");
  if (posMaster.length < 2) {
    console.log("Activate POS Engine: POS_MASTER is empty - nothing to do.");
    return;
  }
  const categoryRulesRaw = readTable("CATEGORY_RULES");
  const activateListRaw = readTable("POS_ACTIVATE_LIST");
  const control = readTable("CONTROL");

  function setting(name: string, fallback: number): number {
    for (let i = 1; i < control.length; i++) {
      if (norm(String(control[i][0])) == norm(name)) {
        const v = Number(control[i][1]);
        return isNaN(v) ? fallback : v;
      }
    }
    return fallback;
  }
  const activateCountByPpt = setting("ACTIVATE_COUNT_BY_PPT", 0);

  let categoryRulesTable: { key: string; value: string }[] = [];
  for (let i = 1; i < categoryRulesRaw.length; i++) {
    categoryRulesTable.push({
      key: norm(String(categoryRulesRaw[i][0])),
      value: norm(String(categoryRulesRaw[i][1])),
    });
  }

  const pmHeaders = (posMaster[0] as string[]).map((h) => String(h));
  const pmIdx = (name: string) => pmHeaders.indexOf(name);
  const posIdCol = pmIdx("posId");
  const statusCol = pmIdx("status");
  const categoryCol = pmIdx("category");
  const overrideTypeCol = pmIdx("managerOverrideType");
  const pptCol = pmIdx("ppt");
  const notesCol = pmIdx("plannerNotes");

  let explicitIds = new Set<string>();
  for (let i = 1; i < activateListRaw.length; i++) {
    const v = String(activateListRaw[i][0] ?? "").trim();
    if (v) {
      explicitIds.add(v);
    }
  }

  const posWs = workbook.getWorksheet("POS_MASTER");
  const today = new Date().toISOString().slice(0, 10);
  let activated: string[] = [];
  let alreadyActive = 0;
  let skippedForceExclude: string[] = [];

  function activateRow(rowIndex: number, row: (string | number | boolean)[], via: string): void {
    const currentOverride = norm(String(row[overrideTypeCol] ?? ""));
    if (currentOverride == "FORCE_INCLUDE") {
      alreadyActive++;
      return;
    }
    posWs.getRangeByIndexes(rowIndex, overrideTypeCol, 1, 1).setValue("FORCE_INCLUDE");
    const existingNotes = String(row[notesCol] ?? "");
    posWs
      .getRangeByIndexes(rowIndex, notesCol, 1, 1)
      .setValue((existingNotes ? existingNotes + " | " : "") + "Hromadně aktivováno" + via + " " + today);
  }

  let mode = "";
  if (explicitIds.size > 0) {
    mode = "seznam (POS_ACTIVATE_LIST)";
    for (let i = 1; i < posMaster.length; i++) {
      const row = posMaster[i];
      const posId = String(row[posIdCol]);
      if (!explicitIds.has(posId)) {
        continue;
      }
      if (String(row[statusCol]) != "Active") {
        continue; // not a candidate at all - Closed POS never enter planning
      }
      const currentOverride = norm(String(row[overrideTypeCol] ?? ""));
      if (currentOverride == "FORCE_EXCLUDE") {
        skippedForceExclude.push(posId); // explicit block always wins - never silently overridden
        continue;
      }
      activateRow(i, row, "");
      activated.push(posId);
    }
  } else if (activateCountByPpt > 0) {
    mode = "prvních " + activateCountByPpt + " podle PPT";
    let pool: { index: number; posId: string; ppt: number; row: (string | number | boolean)[] }[] = [];
    for (let i = 1; i < posMaster.length; i++) {
      const row = posMaster[i];
      if (String(row[statusCol]) != "Active") {
        continue;
      }
      const currentOverride = norm(String(row[overrideTypeCol] ?? ""));
      if (currentOverride == "FORCE_EXCLUDE" || currentOverride == "FORCE_INCLUDE") {
        continue; // already decided one way or the other - not part of the "still excluded" pool
      }
      const category = String(row[categoryCol]);
      const rule = categoryRule(categoryRulesTable, norm(category));
      if (rule != "EXCLUDE") {
        continue; // only pulling from the currently-excluded pool, per the product owner's "vyřazené 1CD" example
      }
      pool.push({ index: i, posId: String(row[posIdCol]), ppt: Number(row[pptCol]) || 0, row });
    }
    pool.sort((a, b) => b.ppt - a.ppt);
    const take = pool.slice(0, activateCountByPpt);
    for (const item of take) {
      activateRow(item.index, item.row, " (PPT)");
      activated.push(item.posId);
    }
  } else {
    console.log(
      "Activate POS Engine: POS_ACTIVATE_LIST is empty and CONTROL.ACTIVATE_COUNT_BY_PPT is 0 - nothing to activate."
    );
    return;
  }

  console.log(
    "Activate POS Engine: mode = " + mode + ". " +
      activated.length + " POS newly/already set to FORCE_INCLUDE (" + alreadyActive + " already were), " +
      skippedForceExclude.length + " skipped (explicit FORCE_EXCLUDE always wins)" +
      (skippedForceExclude.length > 0 ? ": " + skippedForceExclude.join(", ") : "") +
      ". Run Planning Engine next to see them enter this week's plan."
  );
}
